aten/src/ATen/native/CPUBlas.cpp

```diff
@@ -0,0 +1,250 @@
+#include <ATen/native/CPUBlas.h>
+#include <ATen/Config.h>
+
+#include <climits>
+
+#if AT_BUILD_WITH_BLAS()
+extern "C" void dgemm_(char *transa, char *transb, int *m, int *n, int *k, double *alpha, const double *a, int *lda, const double *b, int *ldb, double *beta, double *c, int *ldc);
+extern "C" void sgemm_(char *transa, char *transb, int *m, int *n, int *k, float *alpha, const float *a, int *lda, const float *b, int *ldb, float *beta, float *c, int *ldc);
+extern "C" void cgemm_(char *transa, char *transb, int *m, int *n, int *k, void *alpha, const void *a, int *lda, const void *b, int *ldb, void *beta, void *c, int *ldc);
+extern "C" void zgemm_(char *transa, char *transb, int *m, int *n, int *k, void *alpha, const void *a, int *lda, const void *b, int *ldb, void *beta, void *c, int *ldc);
+#endif  // AT_BUILD_WITH_BLAS()
+
+#ifdef USE_FBGEMM
+#include <fbgemm/FbgemmI64.h>
+#endif  // USE_FBGEMM
+
+namespace at {
+namespace native {
+namespace cpublas {
+namespace {
+
+void normalize_last_dims(
+    TransposeType transa, TransposeType transb,
+    int64_t m, int64_t n, int64_t k,
+    int64_t &lda, int64_t &ldb, int64_t &ldc) {
+  if (n == 1) {
+    ldc = m;
+  }
+
+  if(transa != NoTranspose) {
+    if(m == 1) {
+      lda = k;
+    }
+  } else if(k == 1) {
+    lda = m;
+  }
+
+  if(transb != NoTranspose) {
+    if(k == 1) {
+      ldb = n;
+    }
+  } else if (n == 1) {
+    ldb = k;
+  }
+}
+
+bool use_blas_gemm(
+    TransposeType transa, TransposeType transb,
+    int64_t m, int64_t n, int64_t k,
+    int64_t &lda, int64_t &ldb, int64_t &ldc) {
+  const bool transa_ = transa != NoTranspose;
+  const bool transb_ = transb != NoTranspose;
+  return (
+      (m <= INT_MAX) && (n <= INT_MAX) && (k <= INT_MAX) &&
+      (lda <= INT_MAX) && (ldb <= INT_MAX) && (ldc <= INT_MAX) &&
+      (lda >= std::max(int64_t{1}, (transa_ ? k : m))) &&
+      (ldb >= std::max(int64_t{1}, (transb_ ? n : k))) &&
+      (ldc >= std::max(int64_t{1}, m)));
+}
+
+#if AT_BUILD_WITH_BLAS()
+char to_blas(TransposeType trans) {
+  switch (trans) {
+  case Transpose: return 't';
+  case NoTranspose: return 'n';
+  // case ConjTranspose: return 'c';
```

Comment:
cc @anjali411 we're probably going to want to expose this at some point

Reply:
yeah good point! noted